[projectm-eval] Fix/add "ns-eel2" feature

kblaschke opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/. -->

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

This portfile update fixes installation of the ns-eel2 shim library, adding it as a feature. No upstream version changes in this PR.

Made the feature default on Windows, as this is the only platform the shim is currently useful on, plus most use cases on this platform will require the shim to be available. If the ns-eel2 feature is enabled, it will add an additional package component "Milkdrop", an additional header and a small static library containing the required proxy function implementations.

If this is still considered to violate the "Default features should enable behaviors, not APIs" rule, I can remove it as a default feature, but then most Windows devs would have to enable it manually.

Users of other platforms can still manually enable the feature if required, e.g. when building with MinGW or porting the original Milkdrop code to other platforms.

Also added usage information for the optional "Milkdrop" component added by the "ns-eel2" feature.

JonLiu1993 commented:
Ping @kblaschke for response.

kblaschke commented:
I'm here! Don't see anything to respond to though.

Any issues with the PR that need fixing?

kblaschke commented on ports/projectm-eval/vcpkg.json:
```diff
@@ -13,5 +14,16 @@
     }
-  ]
+  ],
+  "default-features": [
+    {
+      "name": "ns-eel2",
+      "platform": "windows"
+    }
+  ],
```
It's actually an additional static library and header which only makes sense on Windows (and it's the default use case on that platform outside of the projectM library).
If I disable it disable as a default feature I would then have to add a statement to the "usage" file that Windows users should manually enable this feature to be able to use the library within Milkdrop and derivates, which is the major use case for this port and it doesn't work without the feature.

This feature not being there was also a major inconvenience reported by actual users of the port, so I thought making it default would make sense. The guide actually says "should", not "must", so I thought there's some lenience possible with good reasoning. If that's not the case, I'd certainly change it, but would in turn also propose that the maintianer guide is updated to state "Default features _must only_ enable behaviors, not APIs" to make that clear.

JonLiu1993 replied:
Thanks for your explanation.

BillyONeal commented:
If you think most / all users are going to want this feature turned on, what is the value in a user explicitly turning it off? We have found in practice that users find it functionally impossible to disable features marked as defaults given that disabling them only has effects at the top level.

kblaschke commented:
> If you think most / all users are going to want this feature turned on, what is the value in a user explicitly turning it off? We have found in practice that users find it functionally impossible to disable features marked as defaults given that disabling them only has effects at the top level.

Let me explain this again, with a more complete overview on what this feature _is_, what users are using this library, how and whare. Maybe this will make it a bit more clear _why_ I think this feature should be enabled by default on Windows, but not on any other platform:

The library is an expression parser for use in Milkdrop-based audio visualizers, e.g. Milkdrop, Beatdrop, projectM and possibly others as well. The feature in question enabled a small shim that provides a header and some proxy functions to make it easier (basically a drop-in replacement) for the original, assembler-based expression parser that came with the original Milkdrop.

So this library has basically two separate use cases:
1. In projectM, an open-source reimplementation of Milkdrop, which works on basically any platform. I've written the expression parser for projectM, so the API was primarily tailored to fit into this software.
2. In Milkdrop and derived projects like BeatDrop, as a replacement for the original expression parser (called ns_eel2), which is mostly written in x86 32-bit assembly and thus not really portable. The shim feature in discussion adds the header and proxy functions for this exact use case.

Since Milkdrop uses DirectX for rendering, it only works on Windows. There is no port to other platforms or 3D APIs  planned or in progress, and given that Milkdrop's code is hevily based on win32 APIS, there probably never will be any attempt other than a full rewrite (which projectM is).

All that summed up means the feature only makes sense on Windows and will be required by most users on that platform by default, but enabling or disabling it should be possible in either way. Reasoning:

- While it would work just fine, the feature makes no sense on other platforms than Windows, because Milkdrop and its derivates will never be ported to other OSs.
- On Windows, the library will be used by both projectM and any Milkdrop derivates. Since there are more Milkdrop versions out there than projectM, the major use of this library will be in Milkdrop-based applications, requiring the ns-eel2 feature to be _enabled_. For projectM, enabling the feature makes no difference, as it just adds a header and a small sttic lib which just isn't used.
- On other operating systems, there is no Milkdrop, and projectM will (for the foreseeable future) the only software using the projectm-eval library. Other Milkdrop reimplementations deciding to use projectm-eval will most certainly choose to use the main API.

So, I'll repeat myself again here that the maintainer guide uses a softer language by saying:
> Default features _should_ enable behaviors, not APIs 

So I'd interpret this as if there's a good reason to add an API that _only makes sense on a certain platform_ to be enabled as a default feature _on that specific platform_, but still give users the choice of either disabling it on Windows or enabling it on other platforms would certainly not violate the guidelines here.

If the guidelines actually mean, as stated in my other comment, that there is absolutely no room for discussion, then _please_ change the guidelines accordingly to state:
> Default features _must not_ enable behaviors, not APIs 

This will save port contributors and reviewers a lot of unnecessary discussions and frustration.

So, please tell me specifically what I need to do before this PR finally can be merged - either remove the default feature and add proper usage information for the affected devs or keep it as is.

BillyONeal commented:
> All that summed up means the feature only makes sense on Windows and will be required by most users on that platform by default

Then the port can do that. It doesn't need to be a 'feature' in `vcpkg.json`.

> but enabling or disabling it should be possible in either way.

I'm not seeing the scenario where the person on Windows is unhappy if they get this extra feature included with their install.

>I'll repeat myself again here that the maintainer guide uses a softer language by saying:
>>Default features should enable behaviors, not APIs
>So I'd interpret this as if there's a good reason to add an API that only makes sense on a certain platform to be enabled as a default feature on that specific platform, but still give users the choice of either disabling it on Windows or enabling it on other platforms would certainly not violate the guidelines here.

The maintainer guide is kind of unclear about this because we as maintainers have ourselves been kind of unsure what default features should be used for. Over time, we have viewed default features as increasingly broken, because we constantly see people angry that turning off any feature marked as default is next to impossible, because

1. the default way most everyone declares dependencies creates a hard dependency on features declared default (that is, `"dependencies":["projectm-eval"]` in this example)
2. the default features are turned on unless the *top-level* command line or manifest names it. So if you have `projectm/vcpkg.json` that says `"dependencies": [{"name": "projectm-eval", "default-features": false}]` and someone does `vcpkg install projectm`, they *still* get `projectm-eval[ns-eel2]` even though `projectm` said `default-features:false`. They would have to say `vcpkg install project-m projectm-eval[core]`. Or in a manifest they would have to say:
      ```cmake
      "dependencies": [
        "projectm",
        {"name": "projectm-eval", "default-features": false}
      ]
      ```
    even if they only directly only use `project-m` and don't themselves care about projectm-eval.

This has led to constant user feedback that users feel that default features are always forced on. As a result of that feedback, we are viewing adding default features with increasing suspicion, even for the scenarios for which default features were originally designed and as described in the maintainer guide. We have a lot of people who hate how default features work and very few who like how they work.

One of the results of that increasing suspicion is asking questions like "why would a user want to turn this off". Because quite often we have gotten answers like "because a user might be doing Scenario X which requires that the feature be off" and we have to explain the above over again, that making it a default makes it functionally impossible for 99% of users to disable.

>So, please tell me specifically what I need to do before this PR finally can be merged - either remove the default feature and add proper usage information for the affected devs or keep it as is. I could also remove that feature again and just build/install the shim library and header on all platforms, even if it doesn't make sense for users.

What I *want* to say given the scenario you have described above is that I think it should do this:

```cmake
set(BUILD_NS_EEL_SHIM OFF)
if(VCPKG_TARGET_IS_WINDOWS)
    set(BUILD_NS_EEL_SHIM ON)
endif()

vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
    OPTIONS
        -DBUILD_NS_EEL_SHIM=${BUILD_NS_EEL_SHIM}
        -DCMAKE_DISABLE_FIND_PACKAGE_BISON=ON
        -DCMAKE_DISABLE_FIND_PACKAGE_FLEX=ON
)
```

that is, make the portfile just implement the platform specific additions you described, and not create a feature in the first place. But given that I'm not a user of this library myself, I don't understand which users would be made angry by such an outcome, so I'm hesitant to say 'make it look like this'. So I come back to 'why is a user unhappy if this feature is on'.

kblaschke commented:
Thanks for the explanation, that makes a few things clearer!

So, as an alternative, would the following work?
1. Don't expose the shim as a feature on Windows at all, and hard-enable it with your above example.
2. Provide the shim as a feature on all other platforms using `"platform": "!windows"` and thus give users the ability to turn it on when needed, e.g. when compiling with MinGW on Linux.

Since the shim isn't big (under 100 lines of C code, split between header and implementation) and won't add any external dependencies, additional build time or use a lot of disk space, I'd also be fine with just enabling it by default on all platforms and not have it as a feature at all. It'll then be there, but devs can simply ignore it if not needed, as it's available in a separate CMake target/component.

dg0yt commented:
IMO "2." is worse than the current implementation.

"1." still allows the use of the desired part, but it would need to be controlled via triplet file customization (or `--cmake-args`). This might be sufficient for experimental usage. (Fortunately, there are no additional dependencies.)

The last proposal might work as well. Trivial C code doesn't take significant build resources. KISS.

BillyONeal commented:
I would be happy with either (1) (which sounds like exactly my suggestion above) or "just enabling it by default on all platforms" (is that ... 3?). It sounds like nobody would be unhappy for it to not be provided so if it is cheap to do it that makes sense to me.

If someone truly hates that the extra 100 lines of C code are provided they can always use an overlay-port.

kblaschke commented:
Ah, I meant that both 1. and 2. would be part of the suggestion, as 1. is the Windows part and 2. about other platforms, meaning "enable it permanently on Windows and do not provide the feature at all, but on all other platforms provide it as a non-default feature".

I'm personally also tending to the actual second suggestion to always build it on all platforms and it have it as a feature. As you both say, it's really not that much of an overhead and won't hurt anyone, so I'll amend the PR when I'm back at my PC.

Thanks for the good discussion! 👍🏻

kblaschke committed:
[projectm-eval] Fix ns-eel2 shim installation

Previously, the shim's CMake files weren't properly fixed during port installation, making it unusable with CMake.

Also added usage information for the optional "Milkdrop" CMake component.

kblaschke commented:
Replaced the previous changes, only adding the missing CMake fixup for the shim component (which was already built before, but not properly installed) and adding usage information.

This should hopefully fix the discussed problems while providing the shim to users who need it.

BillyONeal commented:
Thanks!

BillyONeal commented:
@kblaschke We updated the documentation in response to your feedback: https://github.com/microsoft/vcpkg-docs/pull/410

kblaschke commented:
Thanks! That change makes it far clearer what the issue with default features actually is and why they should ideally be avoided unless absolutely necessary.
